protocols/kad/src/behaviour.rs

```diff
@@ -732,72 +770,114 @@ where
                                 target = kbucket::Key::new(PeerId::random());
                             }
                             target
-                        }).collect::<Vec<_>>();
+                        }).collect::<Vec<_>>().into_iter();
 
-                    for target in targets {
-                        let info = QueryInfo::Bootstrap { peer: target.clone().into_preimage() };
-                        let peers = self.kbuckets.closest_keys(&target);
-                        let inner = QueryInner::new(info);
-                        self.queries.add_iter_closest(target.clone(), peers, inner);
-                    }
                 }
-                Some(KademliaEvent::BootstrapResult(Ok(BootstrapOk { peer })))
+
+                let num_remaining = remaining.len().checked_sub(1).unwrap_or(0) as u32;
+
+                if let Some(target) = remaining.next() {
+                    let info = QueryInfo::Bootstrap {
+                        peer: target.clone().into_preimage(),
+                        remaining
+                    };
+                    let peers = self.kbuckets.closest_keys(&target);
+                    let inner = QueryInner::new(info);
+                    self.queries.continue_iter_closest(query_id, target.clone(), peers, inner);
+                }
+
+                Some(KademliaEvent::QueryResult {
```

Comment:
Do I understand correctly that instead of a call to `bootstrap` triggering a single `QueryResult` it now triggers one per bootstrap bucket? If so, could you add a note in the changelog?

Reply:
That was already the case before, the only change is in when these queries are started. Previously there were essentially just two stages: 1) start a query to lookup myself 2) start one query per bucket that is further away than the first non-empty bucket. The latter is obviously not so great, apart from the fact that a user doesn't really know when bootstrapping is complete (the events are emitted, but you never know when you got the last one). Now step 2) is made up of N sequential queries, where the query for "refreshing" bucket `n+1` is started only when the query for the "refresh" of bucket `n` has finished.